consensus/polybft/checkpoint_manager.go

```diff
@@ -0,0 +1,271 @@
+		}
+
+		isEndOfEpoch, extra, err := c.isEndOfEpoch(*header, parentHeader)
+		if err != nil {
+			return err
+		}
+
+		parentHeader = header
+
+		// send pending checkpoints only for epoch ending blocks
+		if !isEndOfEpoch {
+			continue
+		}
+
+		err = c.submitCheckpointInternal(nonce, txn, *header, *extra, true)
+		if err != nil {
+			return err
+		}
+
+		nonce++
+	}
+
+	extra, err := GetIbftExtra(latestHeader.ExtraData)
+	if err != nil {
+		return err
+	}
+
+	isEndOfEpoch, _, err := c.isEndOfEpoch(latestHeader, nil)
+	if err != nil {
+		return err
+	}
+
+	return c.submitCheckpointInternal(nonce, txn, latestHeader, *extra, isEndOfEpoch)
+}
+
+// submitCheckpointInternal encodes checkpoint data for the given block and
+// sends a transaction to the CheckpointManager rootchain contract
+func (c *checkpointManager) submitCheckpointInternal(nonce uint64, txn *ethgo.Transaction,
+	header types.Header, extra Extra, isEndOfEpoch bool) error {
+	nextEpochValidators := AccountSet{}
+
+	if isEndOfEpoch {
+		var err error
+		nextEpochValidators, err = c.consensusBackend.GetValidators(header.Number, nil)
+
+		if err != nil {
+			return err
+		}
+	}
+
+	input, err := c.abiEncodeCheckpointBlock(header, extra, nextEpochValidators)
+	if err != nil {
+		return fmt.Errorf("failed to encode checkpoint data to ABI for block %d: %w", header.Number, err)
+	}
+
+	txn.Input = input
+
+	receipt, err := c.rootchain.SendTransaction(nonce, txn)
+	if err != nil {
+		return err
+	}
+
+	if receipt.Status == uint64(types.ReceiptFailed) {
+		return fmt.Errorf("transaction execution failed for block %d", header.Number)
+	}
+
+	return nil
+}
+
+// abiEncodeCheckpointBlock encodes checkpoint data into ABI format for a given header
+func (c *checkpointManager) abiEncodeCheckpointBlock(header types.Header, extra Extra,
+	nextValidators AccountSet) ([]byte, error) {
+	params := map[string]interface{}{
+		"chainID":             new(big.Int).SetUint64(c.blockchain.GetChainID()),
+		"aggregatedSignature": extra.Committed.AggregatedSignature,
+		"validatorsBitmap":    extra.Committed.Bitmap,
+		"epochNumber":         new(big.Int).SetUint64(extra.Checkpoint.EpochNumber),
+		"blockNumber":         new(big.Int).SetUint64(header.Number),
+		"blockHash":           header.Hash,
+		"blockRound":          new(big.Int).SetUint64(extra.Checkpoint.BlockRound),
+		"eventRoot":           extra.Checkpoint.EventRoot.Bytes(),
+		"nextValidators":      nextValidators.AsGenericMaps(),
+	}
+
+	return submitCheckpointMethod.Encode(params)
+}
+
+// setCheckpointsOffset sets new checkpointsOffset value
+func (c *checkpointManager) setCheckpointsOffset(checkpointsOffset uint64) {
+	c.checkpointsOffset = checkpointsOffset
+}
+
+// isCheckpointBlock returns true for epoch ending blocks and
+// blocks in the middle of the epoch which are offseted by predefined count of blocks
+func (c *checkpointManager) isCheckpointBlock(header types.Header) (bool, error) {
+	if header.Number%c.checkpointsOffset == 0 {
```

Comment:
It think its not quite right.
Eg:
start epoch block - 3
epoch offset - 5
epoch size - 20
we need to send checkpoints on 8th, 13th, 18th, 23th blocks.

Reply:
Good catch. Although it seemed to be working good enough when chain is starting from the genesis block... 😕 

In order to make it robust, I'd need to query smart contract to make sure what is the last checkpointed block and to add offset to it and compare it against `header.Number`. Do yo have any better solution?